torch/csrc/jit/script/final_returns.cpp

```diff
@@ -1,106 +1,288 @@
+ * been hit, we add an extra output for the return value, and an extra output
+ * indicating whether or not the return has been hit (a sentinel value).
+ *
+ * When we encounter a node that might return, we guard all subsequent nodes
+ * in the block with the sentinel value of that node.
+ */
+
+// Will a block or node return
+enum Return_Status { WONT_RETURN, MIGHT_RETURN, WILL_RETURN };
+
+// The return output of Control Flow Nodes is the second to last output,
+// sentinel output is the output
+constexpr size_t RETURN_OFFSET = 2;
+constexpr size_t SENTINEL_OFFSET = 1;
 
-void checkNoReturn(const TreeRef& ref) {
-  if (ref->kind() == TK_RETURN)
-    throw ErrorReport(ref) << "return is not allowed from a loop.";
-  for (const TreeRef& child : ref->trees()) {
-    checkNoReturn(child);
+struct EarlyReturns {
+  EarlyReturns(std::shared_ptr<Graph> graph_) : graph(std::move(graph_)){};
+
+  Value* getReturnVal(Node* n) {
+    AT_ASSERT(n->kind() == prim::If || n->kind() == prim::Loop);
+    return n->outputs().at(n->outputs().size() - RETURN_OFFSET);
   }
-}
 
-// transform stmts so that its last action is to return or report that it
-// never returns.
-// return_none - if true, add an implicit `return None` to the end of the block
-//   this handles the case where the return is implicit at the end of the
-//   function.
-ReturnInfo makeReturnsFinal(
-    const SourceRange& range,
-    at::ArrayRef<TreeRef> stmts,
-    bool return_none);
-ReturnInfo makeReturnsFinal(const List<Stmt>& stmts, bool return_none) {
-  return makeReturnsFinal(stmts.range(), stmts.get()->trees(), return_none);
-}
-ReturnInfo makeReturnsFinal(
-    const SourceRange& range,
-    at::ArrayRef<TreeRef> stmts,
-    bool return_none) {
-  std::vector<TreeRef> changed;
-  changed.reserve(stmts.size());
-  for (size_t i = 0; i < stmts.size(); ++i) {
-    const TreeRef& stmt = stmts[i];
-    switch (stmt->kind()) {
-      case TK_IF: {
-        auto if_stmt = If(stmt);
-        auto true_final = makeReturnsFinal(if_stmt.trueBranch(), false);
-        // (3) early return an if statement without an else block:
-        if (true_final.returns_ && if_stmt.falseBranch().size() == 0) {
-          auto rest_final =
-              makeReturnsFinal(range, stmts.slice(i + 1), return_none);
-          if (!rest_final.returns_) {
-            throw ErrorReport(if_stmt)
-                << "This if statement performs an early return, but the block of code that follows it does not return."
-                << " Early returns are only allowed when the block following them also returns.";
-          }
-          changed.emplace_back(
-              if_stmt.withNewBranches(true_final.stmts_, rest_final.stmts_));
-          return {true, List<Stmt>::unsafeCreate(range, std::move(changed))};
-        }
+  Value* getReturnVal(Block* b) {
+    return b->outputs().at(b->outputs().size() - RETURN_OFFSET);
+  }
+
+  Value* getSentinelVal(Node* n) {
```

Comment:
Sentinel is not a great term because in the future there will be more than one sential (maybe one for the function having returned, and one for break/continue). `HasReturned` might be more specific.